game-core/src/main/java/games/strategy/engine/data/GameMap.java

```diff
@@ -271,80 +272,110 @@ public Territory getTerritory(final String s) {
   }
 
   /**
-   * Returns the shortest route between two territories or null if no route exists.
+   * Returns all neighbors within a certain distance of the starting territory that match the
+   * condition. Does NOT include the original/starting territory in the returned Set.
    *
-   * @param t1 start territory of the route
-   * @param t2 end territory of the route
+   * <p>TODO: update to properly consider movement cost not just distance
+   *
+   * @param territory referring territory
+   * @param movementLeft maximal distance of the neighboring territories
    */
-  public Route getRoute(final Territory t1, final Territory t2) {
-    return getRoute(t1, t2, Matches.territoryIsLandOrWater());
+  public Set<Territory> getNeighborsByMovementCost(
+      final Territory territory, final Unit unit, final BigDecimal movementLeft) {
```

Comment:
Why BigDecimal for movement left and not double?

Reply:
I want to avoid precision issues. Movement is kind of like a currency so fits better as a BigDecimal type.

Reply:
Are there cases where you saw precision problems? Will future engineers think this is over-engineered?

Reply:
Could be worth creating a proper abstraction, just like java `Money` class, eg: `MovementPoints`

The below reads well:
```
public Set<Territory> getNeighborsByMovementCost(
      final Territory territory, final Unit unit, final MovementPoints movementLeft)
```

At that rate the internal data representation would be encapsulated and we could change it more easily and/or more easily verify the needed precision. The encapsulation really would help the verification as we'd know all access to/from the data and so we could verify the precision requirements and refactor if needed/desired.

There could be potential for some natural data operations and help avoid the 'compareTo' method of BigDecimal

Reply:
So the double precision problems are fairly common. A very simple example is if you say would want to have road territories which have a 0.2 movement cost (so 1 move unit can move 5 territories). So a simple test that shows the problem would be like this:
```
  void testDouble() {
    final double d1 = 1;
    final double d2 = 0.2;
    final double result = d1 - d2 - d2 - d2 - d2 - d2;
    assertEquals(result, 0);
  }
```

You'll see that it fails as doubles can't handle the precision properly. So we need to do 1 of 3 options of which I think BigDecimal is probably the best:
1. Use BigDecimal
2. Use double/float but round every time you do any comparison of values (painful and very open to developers forgetting to and having edge condition bugs)
3. Create some sort of BigDecimal-like class like Money/MovementPoints which from what we need functionality wise would end up essentially being BigDecimal

So I decided to just go with BigDecimal though option 3 (some wrapper around most likely BigDecimal) could be used. Just didn't see enough value in introducing that complexity.

Reply:
Perhaps better safe than sorry - ok. Though I'm concerned at lack of abstraction, have you put any thought to wrapping the internal data type? The comparisons of BigDecimal seem a bit 'raw' or 'low-level'. For example my misunderstanding of the other if/else chain is a bad sign. I've always thought the idea of a code sitting down with a fine whiskey by a fireplace to study 10 lines of code for an evening is an anti-pattern, not developers being smart (in fact, the 'smart' developer I think is just a myth, there's lots of code to go through, if it does not take little amounts of time to understand, it does not scale).